kratos/tests/strategies/schemes/test_residual_based_adjoint_bossak_scheme.cpp

```diff
@@ -0,0 +1,686 @@
+}
+
+void InitializeAdjointModelPart(ModelPart& rModelPart)
+{
+    rModelPart.AddNodalSolutionStepVariable(DISPLACEMENT);
+    rModelPart.AddNodalSolutionStepVariable(REACTION);
+    rModelPart.AddNodalSolutionStepVariable(VELOCITY);
+    rModelPart.AddNodalSolutionStepVariable(ACCELERATION);
+    rModelPart.AddNodalSolutionStepVariable(ADJOINT_VECTOR_1);
+    rModelPart.AddNodalSolutionStepVariable(ADJOINT_VECTOR_2);
+    rModelPart.AddNodalSolutionStepVariable(ADJOINT_VECTOR_3);
+    rModelPart.AddNodalSolutionStepVariable(AUX_ADJOINT_VECTOR_1);
+    rModelPart.AddNodalSolutionStepVariable(SCALAR_SENSITIVITY);
+    rModelPart.CreateNewNode(1, 0.0, 0.0, 0.0);
+    rModelPart.CreateNewNode(2, 0.0, 0.0, 0.0);
+    rModelPart.SetBufferSize(2);
+    for (auto& r_node : rModelPart.Nodes())
+    {
+        r_node.AddDof(ADJOINT_VECTOR_1_X);
+    }
+    auto p_adjoint_element =
+        AdjointElement::Create(rModelPart.pGetNode(1), rModelPart.pGetNode(2));
+    rModelPart.AddElement(p_adjoint_element);
+}
+
+}
+
+} // unnamed namespace
+
+KRATOS_TEST_CASE_IN_SUITE(ResidualBasedAdjointBossak_TwoMassSpringDamperSystem, KratosCoreSchemesFastSuite)
+{
+    namespace Nlsmd = NonLinearSpringMassDamper;
+    // Solve the primal problem.
+    ModelPart model_part("test");
+    Nlsmd::InitializePrimalModelPart(model_part);
+    auto p_results_data = Kratos::make_shared<Nlsmd::PrimalResults>();
+    Base::PrimalStrategy solver(model_part, p_results_data);
+    solver.Initialize();
+    const double end_time = 0.1;
+    const double start_time = 0.;
+    const std::size_t N = 5;
+    const double delta_time = (end_time - start_time) / N;
+    model_part.CloneTimeStep(start_time - delta_time);
+    model_part.CloneTimeStep(start_time);
+    for (double current_time = start_time; current_time < end_time;)
+    {
+        current_time += delta_time;
+        model_part.CloneTimeStep(current_time);
+        solver.Solve();
+    }
+
+    // Solve the adjoint problem.
+    ModelPart adjoint_model_part("test");
+    Nlsmd::InitializeAdjointModelPart(adjoint_model_part);
+    auto p_response_function =
+        Kratos::make_shared<Nlsmd::ResponseFunction>(adjoint_model_part);
+    Base::AdjointStrategy adjoint_solver(adjoint_model_part, p_results_data, p_response_function);
+    adjoint_solver.Initialize();
+    SensitivityBuilder sensitivity_builder(
+        Parameters{R"({ "element_data_sensitivity_variables": ["SCALAR_SENSITIVITY"] })"},
+        adjoint_model_part, p_response_function);
+    sensitivity_builder.Initialize();
+    adjoint_model_part.CloneTimeStep(end_time + 2. * delta_time);
+    adjoint_model_part.CloneTimeStep(end_time + delta_time);
+    for (double current_time = end_time + delta_time; current_time >= start_time + 1.5 * delta_time;)
+    {
+        current_time -= delta_time;
+        adjoint_model_part.CloneTimeStep(current_time);
+        adjoint_solver.Solve();
+        sensitivity_builder.UpdateSensitivities();
```

Comment:
Just a question: would the current design allow to use the `SensitivityBuilder` as a post processing process of the adjoint analysis?

Reply:
Yes, that should be possible.